artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/ProcessorBase.java

```diff
@@ -108,21 +103,21 @@ public void shutdown(long timeout, TimeUnit unit) {
     * It will wait the current execution (if there is one) to finish
     * but will not complete any further executions
     */
-   public List<T> shutdownNow() {
+   public int shutdownNow(Consumer<? super T> onPendingItem) {
       //alert anyone that has been requested (at least) an immediate shutdown
+      requestedForcedShutdown = true;
       requestedShutdown = true;
-      started = false;
 
       if (inHandler()) {
          stateUpdater.set(this, STATE_FORCED_SHUTDOWN);
       } else {
          //it could take a very long time depending on the current executing task
          do {
             //alert the ExecutorTask (if is running) to just drain the current backlog of tasks
-            final int startState = stateUpdater.get(this);
+            final int startState = this.state;
             if (startState == STATE_FORCED_SHUTDOWN) {
-               //another thread has completed a forced shutdown
-               break;
```

Comment:
Better to exit instead of just stop looping: only one thread is allowed to perform the cleanup/counting of the stopped tasks in order to achieve consistent results.

Reply:
You still need to cleanup down the code.

Reply:
Your change here is wrong, you still need to return the pending items, hence the break.

Reply:
Not sure about this, but probably I've misunderstood the purpose of the method.
Let me explain what I've understood: if 2 concurrent `shutdownNow` are happening only the first one can return the stopped tasks while the other it doesn't need it because has already found the executor in a (final) forced shutdown state, hence it is not responsible of having stopped the executor. Have I misunderstood it?